crates/verify/src/etherscan/mod.rs

```diff
@@ -143,6 +159,302 @@ impl VerificationProvider for EtherscanVerificationProvider {
+
+        // Compare creation code with locally built bytecode and `maybe_creation_code`.
+        let match_type = helpers::match_bytecodes(
+            local_bytecode_vec.as_slice(),
+            maybe_creation_code,
+            &constructor_args,
+            false,
+        );
+
+        let mut json_results: Vec<JsonResult> = vec![];
+        helpers::print_result(
+            &args,
+            match_type,
+            BytecodeType::Creation,
+            &mut json_results,
+            etherscan_metadata,
+            config,
+        );
+
+        // If the creation code does not match, the runtime also won't match. Hence return.
+        if match_type.is_none() {
+            helpers::print_result(
+                &args,
+                None,
+                BytecodeType::Runtime,
+                &mut json_results,
+                etherscan_metadata,
+                config,
+            );
+            if args.json {
+                println!("{}", serde_json::to_string(&json_results)?);
+            }
+            return Ok(());
+        }
+
+        // Get contract creation block.
+        let simulation_block = match args.block {
+            Some(BlockId::Number(BlockNumberOrTag::Number(block))) => block,
+            Some(_) => eyre::bail!("Invalid block number"),
+            None => {
+                let provider = utils::get_provider(&config)?;
+                provider
+                    .get_transaction_by_hash(creation_data.transaction_hash)
+                    .await.or_else(|e| eyre::bail!("Couldn't fetch transaction from RPC: {:?}", e))?.ok_or_else(|| {
+                        eyre::eyre!("Transaction not found for hash {}", creation_data.transaction_hash)
+                    })?
+                    .block_number.ok_or_else(|| {
+                        eyre::eyre!("Failed to get block number of the contract creation tx, specify using the --block flag")
+                    })?
+            }
+        };
+
+        // Fork the chain at `simulation_block`.
+        let (mut fork_config, evm_opts) = config.clone().load_config_and_evm_opts()?;
+        fork_config.fork_block_number = Some(simulation_block - 1);
+        fork_config.evm_version =
+            etherscan_metadata.evm_version()?.unwrap_or(EvmVersion::default());
+        let (mut env, fork, _chain) =
+            TracingExecutor::get_fork_material(&fork_config, evm_opts).await?;
+
+        let mut executor =
+            TracingExecutor::new(env.clone(), fork, Some(fork_config.evm_version), false, false);
+        env.block.number = U256::from(simulation_block);
+        let block = provider.get_block(simulation_block.into(), true.into()).await?;
+
+        // Workaround for the NonceTooHigh issue as we're not simulating prior txs of the same
+        // block.
+        let prev_block_id = BlockId::number(simulation_block - 1);
+        let prev_block_nonce = provider
+            .get_transaction_count(creation_data.contract_creator)
```

Comment:
Would the following work to resolve the discrepancy between Blockscout and Etherscan APIs?

At Blockscout we believe it's more accurate to display the contract creator as an address that directly issued the `CREATE2` opcode, hence the factory address, not the `tx.origin`.

```suggestion
            .get_transaction_count(transaction.from)
```

Reply:
This only resolves it for the case where an EOA calls the create2 factory right? If a contract calls the create2 contract, this reports the EOA as the creator when it really is another contract contract

Reply:
I see. However, you don't replace the transaction sender while re-simulating the tx with altered nonce, right? So the nonce you'll need to fake also should belong to the `transaction.from` instead of address that called `CREATE2` factory. Am I missing something?

Reply:
Can also confirm that etherscan always reports EOA as a contract deployer, even if the factory was called by another contract, e.g. https://etherscan.io/address/0x9a52283276a0ec8740df50bf01b28a80d880eaf2 . I'd assume they simply take `tx.origin` in all the cases. So this won't work as you expect either.

Reply:
> Can also confirm that etherscan always reports EOA as a contract deployer, even if the factory was called by another contract,

Thanks, if this is what etherscan does then I'm ok with doing the same here, cc @zerosnacks